File: matiasmasso/Mat.NET/DTO/DTOUsrLog.cs
using System;

namespace DTO
{
    public class DTOUsrLog
    {
        public DTOUser UsrCreated { get; set; }
        public DTOUser UsrLastEdited { get; set; }
        public DateTime FchCreated { get; set; }
        public DateTime FchLastEdited { get; set; }

        public static DTOUsrLog Factory(DTOUser oUser = null/* TODO Change to default(_) if this is not a reference type */)
        {
            DTOUsrLog retval = new DTOUsrLog();
            {
                var withBlock = retval;
                withBlock.UsrCreated = oUser;
                withBlock.UsrLastEdited = oUser;
            }
            return retval;
        }

        public string text()
        {
            string retval = "";
            var sameUser = UsrCreated != null && UsrCreated.Equals(UsrLastEdited);
            var sameFch = Math.Abs((FchCreated - FchLastEdited).TotalSeconds) < 2;
            var sameUserFch = sameUser & sameFch;
            if (UsrLastEdited == null | sameUserFch)
            {
                if (FchCreated == default(DateTime))
                    retval = string.Format("Registrat per {0}", DTOUser.NicknameOrElse(UsrCreated));
                else
                    retval = string.Format("Registrat per {0} el {1:dd/MM/yy HH:mm}", DTOUser.NicknameOrElse(UsrCreated), FchCreated);
            }
            else if (FchLastEdited == default(DateTime))
                retval = string.Format("Registrat per {0} el {1:dd/MM/yy HH:mm}  i modificat per {2} ", DTOUser.NicknameOrElse(UsrCreated), FchCreated, DTOUser.NicknameOrElse(UsrLastEdited));
            else
                retval = string.Format("Registrat per {0} el {1:dd/MM/yy HH:mm}  i modificat per {2} el {3:dd/MM/yy HH:mm}", DTOUser.NicknameOrElse(UsrCreated), FchCreated, DTOUser.NicknameOrElse(UsrLastEdited), FchLastEdited);
            return retval;
        }

        public static string logText(DTOUser oUserCreated, DTOUser oUserLastEdited = null/* TODO Change to default(_) if this is not a reference type */, DateTime DtFchCreated = default(DateTime), DateTime DtFchLastEdited = default(DateTime))
        {
            var oUsrLog = DTOUsrLog.Factory(oUserCreated);
            {
                var withBlock = oUsrLog;
                withBlock.UsrLastEdited = oUserLastEdited;
                withBlock.FchCreated = DtFchCreated;
                withBlock.FchLastEdited = DtFchLastEdited;
            }
            return oUsrLog.text();
        }

        public DTOTracking Tracking()
        {
            DTOTracking retval = DTOTracking.Factory(DTOCod.Wellknowns.Logged, this.UsrCreated);
            retval.UsrLog.FchCreated = this.FchCreated;
            return retval;
        }
    }
}

File: matiasmasso/Mat.NET/DT2/DTOUsrLog.vb
Public Class DTOUsrLog
    Property usrCreated As DTOUser
    Property usrLastEdited As DTOUser
    Property fchCreated As Date
    Property fchLastEdited As Date

    Shared Function Factory(Optional oUser As DTOUser = Nothing) As DTOUsrLog
        Dim retval As New DTOUsrLog
        With retval
            .UsrCreated = oUser
            .UsrLastEdited = oUser
            '.FchCreated = Now
            '.FchLastEdited = .FchCreated
        End With
        Return retval
    End Function

    Public Function Text() As String
        Dim retval As String = ""
        Dim sameUser = _UsrCreated IsNot Nothing AndAlso _UsrCreated.Equals(_UsrLastEdited)
        Dim sameFch = Math.Abs((_FchCreated - FchLastEdited).TotalSeconds) < 2
        Dim sameUserFch = sameUser And sameFch
        If _UsrLastEdited Is Nothing Or sameUserFch Then
            If _FchCreated = Nothing Then
                retval = String.Format("Registrat per {0}", DTOUser.NicknameOrElse(_UsrCreated))
            Else
                retval = String.Format("Registrat per {0} el {1:dd/MM/yy HH:mm}", DTOUser.NicknameOrElse(_UsrCreated), _FchCreated)
            End If
        Else
            If _FchLastEdited = Nothing Then

                retval = String.Format("Registrat per {0} el {1:dd/MM/yy HH:mm}  i modificat per {2} ", DTOUser.NicknameOrElse(_UsrCreated), _FchCreated, DTOUser.NicknameOrElse(_UsrLastEdited))
            Else
                retval = String.Format("Registrat per {0} el {1:dd/MM/yy HH:mm}  i modificat per {2} el {3:dd/MM/yy HH:mm}", DTOUser.NicknameOrElse(_UsrCreated), _FchCreated, DTOUser.NicknameOrElse(_UsrLastEdited), FchLastEdited)
            End If
        End If
        Return retval
    End Function

    Shared Function LogText(oUserCreated As DTOUser, 'Per compatibilitat amb antics
                            Optional oUserLastEdited As DTOUser = Nothing,
                            Optional DtFchCreated As Date = Nothing,
                            Optional DtFchLastEdited As Date = Nothing) As String

        Dim oUsrLog = DTOUsrLog.Factory(oUserCreated)
        With oUsrLog
            .UsrLastEdited = oUserLastEdited
            .FchCreated = DtFchCreated
            .FchLastEdited = DtFchLastEdited
        End With
        Return oUsrLog.Text
    End Function


End Class
